rules/DeadCode/Rector/FunctionLike/TooWideReturnTypeRector.php

```diff
@@ -0,0 +1,243 @@
+            return true;
+        }
+
+        if (! $node instanceof ClassMethod) {
+            return false;
+        }
+
+        if ($node->isPrivate() || $node->isFinal()) {
+            return false;
+        }
+
+        if ($node->isAbstract()) {
+            return true;
+        }
+
+        $classReflection = $this->reflectionResolver->resolveClassReflection($node);
+
+        if (! $classReflection instanceof ClassReflection) {
+            return true;
+        }
+
+        return ! $classReflection->isFinal();
+    }
+
+    /**
+     * @param Return_[] $returnStatements
+     * @return Type[]
+     */
+    private function collectActualReturnTypes(
+        ClassMethod|Function_|Closure|ArrowFunction $node,
+        array $returnStatements,
+        bool $isAlwaysTerminating,
+    ): array {
+        if ($node instanceof ArrowFunction) {
+            return [$this->getType($node->expr)];
+        }
+
+        $returnTypes = [];
+        foreach ($returnStatements as $returnStatement) {
+            if ($returnStatement->expr === null) {
+                $returnTypes[] = new NullType();
+                continue;
+            }
+
+            $returnTypes[] = $this->getType($returnStatement->expr);
+        }
+
+        if (! $isAlwaysTerminating) {
+            $returnTypes[] = new NullType();
+        }
+
+        return $returnTypes;
+    }
+
+    /**
+     * @param Type[] $actualReturnTypes
+     */
+    private function narrowUnionReturnType(
+        UnionType $unionType,
+        array $actualReturnTypes
+    ): ComplexType|Identifier|Name|null {
+        $types = $unionType->types;
+        $usedTypes = [];
+
+        foreach ($types as $type) {
+            $declaredType = $this->getType($type);
+            if ($declaredType instanceof MixedType) {
+                // Mixed type covers all other types, so we should only keep mixed
+                return new Identifier('mixed');
+            }
+            foreach ($actualReturnTypes as $actualType) {
+                if (! $declaredType->isSuperTypeOf($actualType)->no()) {
+                    $usedTypes[] = $declaredType;
+                    break;
+                }
+            }
+        }
+
+        if ($usedTypes === [] || count($usedTypes) === count($types)) {
+            return null;
+        }
+
+        return $this->staticTypeMapper->mapPHPStanTypeToPhpParserNode(
+            TypeCombinator::union(...$usedTypes),
+            TypeKind::RETURN
+        );
+    }
+
+    private function hasYield(ClassMethod|Function_|Closure|ArrowFunction $node): bool
+    {
+        if ($node instanceof ArrowFunction) {
+            return false;
+        }
+
+        $stmts = $node->stmts;
+
+        if ($stmts === null || $stmts === []) {
+            return false;
+        }
+
+        return (bool) $this->betterNodeFinder->findFirst(
+            $stmts,
+            fn (Node $subNode): bool => $subNode instanceof Yield_ || $subNode instanceof YieldFrom
+        );
```

Comment:
use only scoped search: 

```suggestion
        return (bool) $this->betterNodeFinder->hasInstancesOfInFunctionLikeScoped($node, [Yield_::class, YieldFrom::class]);